packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -126,14 +134,30 @@ export default class WorkspacesForeachCommand extends BaseCommand {
       ? project.topLevelWorkspace
       : cwdWorkspace!;
 
+    if (configuration.projectCwd === null)
+      throw new UsageError(`This command can only be run from within a Yarn project`);
+
+    const root = this.since ? await gitUtils.fetchRoot(configuration.projectCwd) : null;
+    const base = this.since && root !== null
+      ? await gitUtils.fetchBase(root, {baseRefs: typeof this.since === `string` ? [this.since] : configuration.get(`changesetBaseRefs`)})
+      : null;
```

Comment:
I think this is a good idea. I can pull this out into a new function in gitUtils if there are no objections to it living there.

My only hesitation was around calling `changesetBaseRefs` config value from `@yarnpkg/plugin-git` as it seems that was originally intended to be used just by `@yarnpkg/plugin-version`, so I didn't want to proliferate it too widely. But I guess we are already using it in `@yarnpkg/plugin-essentials` and `@yarnpkg/plugin-workspace-tools` so there shouldn't be much harm.

I think the tests will be reliable enough to detect issues here though. I plan on running this command both with `--since` as true and where `--since=<ref>` is set to a specific ref.

Reply:
Fixed in 510516aeafc0dd8438cf1bfa2d55aed63415502f.
Moved all this into `fetchChangedWorkspaces`.